File: backend/src/main/java/com/snlbanking/service/UserService.java
package com.snlbanking.service;

import com.snlbanking.dto.UserRegistrationDTO;
import com.snlbanking.dto.UserResponseDTO;
import com.snlbanking.entity.User;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * User Service Interface
 * Defines business logic operations for User entity
 * 
 * @author SNL Banking Team
 * @version 1.0.0
 */
public interface UserService {

    /**
     * Register a new user
     */
    UserResponseDTO registerUser(UserRegistrationDTO registrationDTO);

    /**
     * Get user by ID
     */
    UserResponseDTO getUserById(Long userId);

    /**
     * Get user by email
     */
    User getUserByEmail(String email);

    /**
     * Update user profile
     */
    UserResponseDTO updateUserProfile(Long userId, UserRegistrationDTO updateDTO);

    /**
     * Get all users with pagination
     */
    Page<UserResponseDTO> getAllUsers(Pageable pageable);

    /**
     * Search users by name
     */
    Page<UserResponseDTO> searchUsersByName(String name, Pageable pageable);

    /**
     * Get users by role
     */
    Page<UserResponseDTO> getUsersByRole(User.UserRole role, Pageable pageable);

    /**
     * Get users by status
     */
    Page<UserResponseDTO> getUsersByStatus(User.UserStatus status, Pageable pageable);

    /**
     * Change user password
     */
    void changePassword(Long userId, String oldPassword, String newPassword);

    /**
     * Reset user password
     */
    void resetPassword(String email);

    /**
     * Activate user account
     */
    void activateUser(Long userId);

    /**
     * Deactivate user account
     */
    void deactivateUser(Long userId);

    /**
     * Delete user (soft delete)
     */
    void deleteUser(Long userId);
    // Add this method to your existing UserService interface
UserResponseDTO login(String email, String password);
}

File: frontend/src/services/userService.js
import apiClient from './apiClient';
import { USER_ENDPOINTS } from '../constants/apiEndpoints';

const userService = {
  // Get user by ID
  getUserById: async (userId) => {
    try {
      const response = await apiClient.get(USER_ENDPOINTS.GET_USER(userId));
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Get all users (Admin)
  getAllUsers: async (page = 0, size = 10) => {
    try {
      const response = await apiClient.get(USER_ENDPOINTS.GET_ALL_USERS, {
        params: { page, size },
      });
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Update user profile
  updateUserProfile: async (userId, userData) => {
    try {
      const response = await apiClient.put(USER_ENDPOINTS.UPDATE_USER(userId), userData);
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Activate user
  activateUser: async (userId) => {
    try {
      const response = await apiClient.post(USER_ENDPOINTS.ACTIVATE_USER(userId));
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Deactivate user
  deactivateUser: async (userId) => {
    try {
      const response = await apiClient.post(USER_ENDPOINTS.DEACTIVATE_USER(userId));
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Delete user
  deleteUser: async (userId) => {
    try {
      const response = await apiClient.delete(USER_ENDPOINTS.DELETE_USER(userId));
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Search users by name
  searchUsers: async (name) => {
    try {
      const response = await apiClient.get(USER_ENDPOINTS.SEARCH_USERS, {
        params: { name },
      });
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },
};

export default userService;
